src/bindings/js/node/src/async_infer_queue.cpp

```diff
@@ -0,0 +1,209 @@
+                           InstanceMethod("startAsync", &AsyncInferQueue::start_async),
+                           InstanceMethod("release", &AsyncInferQueue::release),
+                       });
+}
+
+int AsyncInferQueue::check_idle_request_id() {
+    std::unique_lock<std::mutex> lock(m_mutex);
+    if (m_idle_handles.empty()) {
+        return -1;
+    }
+    int idle_handle = static_cast<int>(m_idle_handles.front());
+    // wait for request to make sure it returned from callback
+    m_requests[idle_handle].wait();
+    m_idle_handles.pop();
+    return idle_handle;
+}
+
+void AsyncInferQueue::set_default_callbacks() {
+    for (size_t handle = 0; handle < m_requests.size(); handle++) {
+        m_requests[handle].set_callback([this, handle](std::exception_ptr exception_ptr) {
+            {
+                std::lock_guard<std::mutex> lock(m_mutex);
+                m_idle_handles.push(handle);
+            }
+            try {
+                if (exception_ptr) {
+                    std::rethrow_exception(exception_ptr);
+                }
+            } catch (const std::exception& e) {
+                OPENVINO_THROW(e.what());
+            }
+        });
+    }
+}
+
+void AsyncInferQueue::set_custom_callbacks(const Napi::CallbackInfo& info) {
+    std::vector<std::string> allowed_signatures;
+    try {
+        OPENVINO_ASSERT(ov::js::validate<Napi::Function>(info, allowed_signatures),
+                        "'set_callback'",
+                        ov::js::get_parameters_error_msg(info, allowed_signatures));
+
+        m_tsfn =
+            Napi::ThreadSafeFunction::New(info.Env(), info[0].As<Napi::Function>(), "AsyncInferQueueCallback", 0, 1);
+        for (size_t handle = 0; handle < m_requests.size(); handle++) {
+            m_requests[handle].set_callback([this, handle](std::exception_ptr exception_ptr) {
+                if (exception_ptr == nullptr) {
+                    auto ov_callback = [this](Napi::Env env, Napi::Function user_callback, int* handle) {
+                        Napi::Object js_ir = InferRequestWrap::wrap(env, m_requests[*handle]);
+                        const auto promise = m_user_ids[*handle].second;
+                        try {
+                            user_callback.Call({js_ir, m_user_ids[*handle].first.Value()});
+                            promise.Resolve(m_user_ids[*handle].first.Value());
+                            // returns before the promise's .then() is completed
+                        } catch (Napi::Error& e) {
+                            promise.Reject(Napi::Error::New(env, e.Message()).Value());
+                        }
+                        {
+                            // Start async inference on the next request or add idle handle to queue
+                            std::lock_guard<std::mutex> lock(m_mutex);
+                            if (m_awaiting_requests.size() > 0) {
```

Comment:
```suggestion
                        if (std::lock_guard<std::mutex> lock(m_mutex); m_awaiting_requests.size() > 0) {
```
Then the whole `if...else if..else` block will have local scope guard and additional code block can be removed.